neuroptica/nonlinearities.py

```diff
@@ -402,3 +405,156 @@ def df_dRe(self, a: np.ndarray, b: np.ndarray) -> np.ndarray:
 
     def df_dIm(self, a: np.ndarray, b: np.ndarray) -> np.ndarray:
         return (a > 0) * (b > 0) * 1j
+
+class TrainableNonLinearity(ComplexNonlinearity):
+    '''
+    Nonlinearity with a trainable parameter
+    '''
+
+    def __init__(self, N, holomorphic, mode):
+        super().__init__(N, holomorphic, mode)
+
+    def compute_gradients(self, inputs: np.ndarray, delta:np.ndarray):
+        '''
+        Compute the gradients for nonlinearity in this layer, without adjusting the parameters
+        :param forward_field: forward-propagating input electric field at the beginning of the nonlinearity
+        :param adjoint_field: backward-propagating output electric field at the end of the nonlinearity
+        :return:
+        '''
+        raise NotImplementedError
+
+
+class FabryPerotInferometer(TrainableNonLinearity):
+    '''
+    CMOS light detector actuated FBI
+    '''
+
+    def __init__(self, N, KAPPA, INIT_VAR):
+        super().__init__(N, holomorphic=False, mode="full")
+
+        self.KAPPA = KAPPA
+        self.N = N
+
+        self.X0_re: np.ndarray = np.random.uniform(size=(N,1), low = -INIT_VAR, high=INIT_VAR)
+        self.X0_im: np.ndarray = np.random.uniform(size=(N,1), low = -INIT_VAR, high=INIT_VAR)
+
+    def forward_pass(self, X: np.ndarray):
+        X_re = np.real(X)
+        X_im = np.imag(X)
+        return (0.5*self.KAPPA*X)/((X_re**2 + X_im**2 - self.X0_re**2 - self.X0_im**2)*-1j + 0.5*self.KAPPA)
+
+    def dRe_dRe(self, x: np.ndarray, y: np.ndarray):
```

Comment:
please compactify the operations here so it is more clear what you are doing. (e.g. get rid of extraneous helper variables - you can just do `x**2` instead of `x2`, `(0.5 * self.KAPPA)**2` instead of `half_k2`, etc.)